File: go/algorithms/dfs.go
package algorithms

// DFS performs depth-first search traversal.
// Time: O(V + E)
// Space: O(V)
func DFS[T comparable](graph *Graph[T], start T, callback func(T)) []T {
	result := []T{}

	if !graph.HasVertex(start) {
		return result
	}

	visited := make(map[T]bool)
	stack := []T{start}

	for len(stack) > 0 {
		current := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		if visited[current] {
			continue
		}

		visited[current] = true
		result = append(result, current)

		if callback != nil {
			callback(current)
		}

		neighbors := graph.Neighbors(current)
		for i := len(neighbors) - 1; i >= 0; i-- {
			if !visited[neighbors[i]] {
				stack = append(stack, neighbors[i])
			}
		}
	}

	return result
}

// DFSRecursive performs recursive depth-first search traversal.
func DFSRecursive[T comparable](graph *Graph[T], start T, callback func(T)) []T {
	result := []T{}

	if !graph.HasVertex(start) {
		return result
	}

	visited := make(map[T]bool)
	dfsRecursive(graph, start, visited, &result, callback)
	return result
}

func dfsRecursive[T comparable](graph *Graph[T], current T, visited map[T]bool, result *[]T, callback func(T)) {
	visited[current] = true
	*result = append(*result, current)

	if callback != nil {
		callback(current)
	}

	for _, neighbor := range graph.Neighbors(current) {
		if !visited[neighbor] {
			dfsRecursive(graph, neighbor, visited, result, callback)
		}
	}
}

// DFSFindPath finds a path between two vertices using DFS.
func DFSFindPath[T comparable](graph *Graph[T], start, end T) []T {
	if !graph.HasVertex(start) || !graph.HasVertex(end) {
		return nil
	}

	if start == end {
		return []T{start}
	}

	visited := make(map[T]bool)
	path := []T{}

	if dfsPath(graph, start, end, visited, &path) {
		return path
	}
	return nil
}

func dfsPath[T comparable](graph *Graph[T], current, end T, visited map[T]bool, path *[]T) bool {
	visited[current] = true
	*path = append(*path, current)

	if current == end {
		return true
	}

	for _, neighbor := range graph.Neighbors(current) {
		if !visited[neighbor] {
			if dfsPath(graph, neighbor, end, visited, path) {
				return true
			}
		}
	}

	*path = (*path)[:len(*path)-1]
	return false
}

// HasCycle detects if a directed graph has a cycle starting from start.
func HasCycle[T comparable](graph *Graph[T], start T) bool {
	if !graph.HasVertex(start) {
		return false
	}

	visited := make(map[T]bool)
	recStack := make(map[T]bool)

	return detectCycle(graph, start, visited, recStack)
}

func detectCycle[T comparable](graph *Graph[T], current T, visited, recStack map[T]bool) bool {
	visited[current] = true
	recStack[current] = true

	for _, neighbor := range graph.Neighbors(current) {
		if !visited[neighbor] {
			if detectCycle(graph, neighbor, visited, recStack) {
				return true
			}
		} else if recStack[neighbor] {
			return true
		}
	}

	recStack[current] = false
	return false
}

File: typescript/src/algorithms/dfs.ts
/**
 * Depth-First Search implementation.
 * Time: O(V + E)
 * Space: O(V)
 */

import { Graph } from './graph';

export function dfs<T>(graph: Graph<T>, start: T, callback?: (vertex: T) => void): T[] {
  const result: T[] = [];

  if (!graph.hasVertex(start)) {
    return result;
  }

  const visited = new Set<T>();
  const stack: T[] = [start];

  while (stack.length > 0) {
    const current = stack.pop()!;

    if (visited.has(current)) {
      continue;
    }

    visited.add(current);
    result.push(current);

    if (callback) {
      callback(current);
    }

    const neighbors = graph.neighbors(current);
    for (let i = neighbors.length - 1; i >= 0; i--) {
      if (!visited.has(neighbors[i])) {
        stack.push(neighbors[i]);
      }
    }
  }

  return result;
}

export function dfsRecursive<T>(graph: Graph<T>, start: T, callback?: (vertex: T) => void): T[] {
  const result: T[] = [];

  if (!graph.hasVertex(start)) {
    return result;
  }

  const visited = new Set<T>();
  dfsRecursiveHelper(graph, start, visited, result, callback);
  return result;
}

function dfsRecursiveHelper<T>(
  graph: Graph<T>,
  current: T,
  visited: Set<T>,
  result: T[],
  callback?: (vertex: T) => void
): void {
  visited.add(current);
  result.push(current);

  if (callback) {
    callback(current);
  }

  for (const neighbor of graph.neighbors(current)) {
    if (!visited.has(neighbor)) {
      dfsRecursiveHelper(graph, neighbor, visited, result, callback);
    }
  }
}

export function dfsFindPath<T>(graph: Graph<T>, start: T, end: T): T[] | undefined {
  if (!graph.hasVertex(start) || !graph.hasVertex(end)) {
    return undefined;
  }

  if (start === end) {
    return [start];
  }

  const visited = new Set<T>();
  const path: T[] = [];

  if (dfsPathHelper(graph, start, end, visited, path)) {
    return path;
  }
  return undefined;
}

function dfsPathHelper<T>(
  graph: Graph<T>,
  current: T,
  end: T,
  visited: Set<T>,
  path: T[]
): boolean {
  visited.add(current);
  path.push(current);

  if (current === end) {
    return true;
  }

  for (const neighbor of graph.neighbors(current)) {
    if (!visited.has(neighbor)) {
      if (dfsPathHelper(graph, neighbor, end, visited, path)) {
        return true;
      }
    }
  }

  path.pop();
  return false;
}

export function hasCycle<T>(graph: Graph<T>, start: T): boolean {
  if (!graph.hasVertex(start)) {
    return false;
  }

  const visited = new Set<T>();
  const recStack = new Set<T>();

  return detectCycle(graph, start, visited, recStack);
}

function detectCycle<T>(
  graph: Graph<T>,
  current: T,
  visited: Set<T>,
  recStack: Set<T>
): boolean {
  visited.add(current);
  recStack.add(current);

  for (const neighbor of graph.neighbors(current)) {
    if (!visited.has(neighbor)) {
      if (detectCycle(graph, neighbor, visited, recStack)) {
        return true;
      }
    } else if (recStack.has(neighbor)) {
      return true;
    }
  }

  recStack.delete(current);
  return false;
}
